Rover: improved vectored thrust support

rmackay9 opened the pull request:
This improves Rover's support for vectored thrust vehicles ([wiki](https://ardupilot.org/rover/docs/rover-vectored-thrust.html)) like boats in which the motor is aimed using the steering servo.

This PR adds:

- SITL gets a new "rover-vectored" model which more accurately represents how these vehicles respond to throttle and steering inputs
- Rover's AP_MotorsUGV class's "MOT_VEC_THR_BASE" parameter is replaced by "MOT_VEC_ANGLEMAX" which allows the user to specify the motor's maximum angle of deflection (in degrees) from the center (i.e. the motor direction's angle from center when the servo is at it's maximum).  This is used as part of the calculation to turn the motor library steering & throttle inputs into the best outputs.

This change is in response to some [real-world testing I did with my DeSET boat](https://www.youtube.com/watch?v=pEwL4FO5bQk).

This has been tested in SITL but not yet on a real vehicle.  In SITL it seems to work well including driving the virtual boat in reverse around a course.

One slightly odd thing I noticed was that the ATC_STR_RAT0_FF value (in SITL) needs to be really small (0.04 vs the regular 0.2).

----------------

The calculations used in SITL are derived from this web page: https://keisan.casio.com/exec/system/1223014690

![randy-rover-vectored-sitl-calc](https://user-images.githubusercontent.com/1498098/113141214-61cad600-9264-11eb-8703-bb203a354869.png)

The SITL model gets two inputs (steering and throttle) that start as PWM inputs but are soon converted to normalised values in the range -1 to +1.

- the steering input is easily converted to an angle (see "steering_angle_rad") by multiplying it by the "vectored_angle_max" (90deg) which is how far the virtual motor can be turned
- this angle is then used to break the incoming throttle value into steering and forward thrust components which are then written back into "steering" and "throttle".
- the "calc_yaw_rate()" method is modified so its result is the above "steering" value multiplied by the "vectored_turn_rate_max" constant.  This essentially means if the steering is turned fully to 90degrees and full throttle is applies the vehicle will rotate at "vectored_turn_rate_max" (currently set to 90deg/sec).

srmainwaring commented:
@rmackay9 if you are interested I’d be happy to test this out in Gazebo / SITL. If you have a STL file for the boat hull and dimensions / mass I could model your DeSET boat and drive, or alternatively use a variant of a model I’ve put together for another AP user who wanted a script for a set up with a pair of reversible jet drives on a bait boat. Gazebo env can model waves and wind drag on the hull.

rmackay9 commented:
@srmainwaring,

I'm very interested in getting help with testing!  I don't have a SITL file for the DeSET boat but any boat with thrusters are rotated with servos should be a good test.  The only limitation is that this only support the motors being aimed at up to 90deg.

IamPete1 commented on Rover/AP_MotorsUGV.cpp:
```diff
@@ -613,9 +615,52 @@ void AP_MotorsUGV::output_regular(bool armed, float ground_speed, float steering
+                    float steering_angle_rad = tanf(steering_norm / throttle_norm);
+                    const float vector_angle_max_rad = radians(constrain_float(_vector_angle_max, 0.0f, 90.0f));
+                    // limit steering angle to vector_angle_max
+                    if (fabsf(steering_angle_rad) > fabsf(vector_angle_max_rad)) {
```
```suggestion
                    if (fabsf(steering_angle_rad) > vector_angle_max_rad) {
```
Unneeded abs

rmackay9 replied:
fixed, thanks.

IamPete1 commented on Rover/AP_MotorsUGV.cpp:
```diff
@@ -613,9 +615,52 @@ void AP_MotorsUGV::output_regular(bool armed, float ground_speed, float steering
+                    const float vector_angle_max_rad = radians(constrain_float(_vector_angle_max, 0.0f, 90.0f));
+                    // limit steering angle to vector_angle_max
+                    if (fabsf(steering_angle_rad) > fabsf(vector_angle_max_rad)) {
+                        steering_angle_rad = constrain_float(steering_angle_rad, -fabsf(vector_angle_max_rad), fabsf(vector_angle_max_rad));
```
```suggestion
                        steering_angle_rad = constrain_float(steering_angle_rad, -vector_angle_max_rad, vector_angle_max_rad);
```

rmackay9 replied:
fixed, txs!

rmackay9 commented:
@srmainwaring, awesome, thanks a lot for your help!

IamPete1 reviewed (commented):
This is better than what we have currently. 

In the case of a 'outboard motor' with a rudder surface, like the picture on the wiki, it can still act as a rudder with no throttle. I think this will not work that well, it would need speed scaling, and you can still steer with no throttle, for example in the case where your going fast and want to stop so set the throttle to 0 and that also constrains your steering to 0. 

Maybe its just a matter of making it clear what type of vehicle this is for and what it is not for.

rmackay9 commented:
@IamPete1,

Txs very much for the review and feedback.  I thought you might be the best one to have a look at this change so thanks!

I agree that it's not great for a motor which also has a fin on it.  The old method didn't scale the steering with speed either but this change has the potential issue that it will straighten out the servo when the throttle goes to zero.  Also because throttle+steering and intimately connected, if the throttle controller is very badly tuned this could lead to the steering servo wagging a loy.  Something to keep an eye out for during testing and should possibly be documented on the wiki.

I agree we should at least change the picture on the wiki to make it clear that it's meant for boats with vectored thrust and no fin.

rmackay9 commented on Rover/AP_MotorsUGV.cpp:
```diff
@@ -613,9 +615,52 @@ void AP_MotorsUGV::output_regular(bool armed, float ground_speed, float steering
+                const float throttle_norm = throttle / 1000.0f;
+
+                // steering can never be more than throttle
+                if (fabsf(steering_norm) > fabsf(throttle_norm)) {
```
Ah, I've realised that this limitation that, "steering can never be more than throttle" is only correct if the MOT_VEC_ANGLEMAX is 45deg (i.e. the steering can be turned 45deg).  The limitation is actually lower if the angle is less then 45deg and higher if > 45deg.  I'll have to fix this.

srmainwaring commented:
@rmackay9 I've got a basic model up and running on this branch using a Lua script to mix the throttle output to a jet drive and reverse servo. I found with `MOT_VEC_ANGLEMAX = 30` it was very hard to get a good tune. That may be because of poor throttle tuning which you mentioned above.

One reason of the need for a small FF value might be that, in simulation at least, the steering rate response is quite non-linear with RC input. At constant throttle (say 50%) the steering response is approx linear for about 0 - 5 deg then very quickly the angular rate increases rapidly with small increases in thrust deflection until the boat locks into a spin. I'll try to get the tune stabilised with 'MOT_VEC_ANGLEMAX = 0' then re-enable with it active.

rmackay9 commented:
@srmainwaring,

Really cool simulation and I love the video!  Make sure the MOT_VEC_ANGLEMAX parameter matches the actual angle of defection of the thruster achieved when the appropriate SERVOn output reaches SERVOn_MIN or SERVOn_MAX.

srmainwaring commented on Rover/AP_MotorsUGV.cpp:
```diff
@@ -613,9 +615,52 @@ void AP_MotorsUGV::output_regular(bool armed, float ground_speed, float steering
+                    // convert steering angle to steering output
+                    steering = steering_angle_rad / vector_angle_max_rad * 4500.0f;
+
+                    // scale up throttle to compensate for steering angle
+                    const float throttle_scaler_inv = cosf(steering_angle_rad);
+                    if (!is_zero(throttle_scaler_inv)) {
+                        throttle /= throttle_scaler_inv;
+                        if (throttle >= 100.0f) {
+                            limit.throttle_upper = true;
+                        }
+                        if (throttle <= -100.0f) {
+                            limit.throttle_lower = true;
+                        }
+                    }
```
Scaling the throttle up to preserve the longitudinal component of thrust will scale up the lateral component as well - which will amplify the yaw effect as the steering angle is increased.

srmainwaring commented:
@rmackay9 I've been running some tests on my simulation in manual mode to get a feel for the behaviour. Now I'm not sure how closely the sim physics corresponds to a real boat, so some of the observations may be off but this is what I see for the jet drive:

The turn rate (as measured with GyroZ) is to a good approximation linear with the thrust angle. This makes sense as sin(x) ~ x for x << 1.

![rover_vec_thrust_steer_rate_1](https://user-images.githubusercontent.com/24916364/113349282-b424f900-932f-11eb-9dfe-61288150cbce.png)

There is a very sudden breakdown in behaviour above approx 6.5 deg thrust angle when the lateral acceleration exceeds 0.18g. At this point the lateral resistance is insufficient to stop the vehicle from spinning, and forward motion ceases.
 
![rover_vec_thrust_steer_rate_2](https://user-images.githubusercontent.com/24916364/113349350-cdc64080-932f-11eb-9d68-d9a36f9a720b.png)

This may be a simulation artefact, or the result of the hull shape used. The boat sim uses a velocity dependent drag and pressure force to prevent side slip - at some point this is overcome by the lateral force. Interesting I've observed similar behaviour in a simulation of Peter's fan car which relies on Coulomb friction forces- but I've not carried out the same analysis yet for a proper comparison.   

What I do notice is that in the valid regime the vehicle velocity is unaffected by the (small) steering angle: i.e. holding RC3 (throttle) constant, the ground speed remains constant as the turn rate is increased, up until the lateral resistance fails. In my simulation at least it may not be necessary to scale up the throttle to compensate for the steering angle (but this may be because of the small angles involved).

rmackay9 commented:
@srmainwaring, thanks again for the analysis.  It's very interesting that the boat suddenly fails like that but 0.18G is a pretty large lateral acceleration for a boat.  I wonder if a model of a less powerful boat with an outboard engine might be a better match for this feature.

By the way, I've found that in SITL the boat handles very badly if the maximum angle is increased to 90deg (45deg works fine) but I'm unsure where between 45 and 90deg is starts to go wrong.  I'll look for a bug in the code.

srmainwaring commented:
@rmackay9 for steering angles > 45 deg you might be seeing the scaling affect on the lateral thrust resulting from applying the throttle_scale_inv factor. For these angles the lateral component gets magnified significantly.

rmackay9 commented:
@srmainwaring, txs.. but that shouldn't be the case unless I've made a mistake in the calculations.  The way it should work is it takes the incoming throttle (which is really a desired longitudinal thrust) and the steering (which is interpreted as a percentage of the maximum possible lateral thrust the vehicle could produce when the steering is at _vector_angle_max with full throttle). These two are the vertical and horizontal lengths of a right-angle triangle.  Then the throttle is scaled up to become the hypotenuse.

srmainwaring commented:
@rmackay9 you're right - I was looking at results for constant steering angle. I'll update my Python worksheet using the desired steering rate and throttle as inputs.

The steering angle calculation should use the arc tangent to get the angle from the triangle side lengths (btw the current calc would give the wrong angle but also nasty results at steering_norm / throttle_norm = pi/2 which a PID controller couldn't adjust for):

```diff
- float steering_angle_rad = tanf(steering_norm / throttle_norm);
+ float steering_angle_rad = arctanf(steering_norm / throttle_norm);
```

Is the scaling of the normalised throttle correct? `throttle_norm` is calculated by dividing through by 1000.0f, but lower down the scaling calculation using `throttle_scaler_inv` clamps the throttle between -100.0f and 100.f. (I'm still getting familiar with the various scaling conventions used in ArduPilot).

rmackay9 commented:
@srmainwaring,

Wow, fantastic testing and fixes!  I'll double check this and then hopefully update the code!

rmackay9 commented:
@srmainwaring,

You were absolutely right!

I've updated this branch with your two fixes and two more I found related to how steering was being limited by throttle and also the steering limits flags didn't seem to be working.  With all these improvements I think it's working quite well now.

Do you want to give it a final check?

P.S. I added you as co-author on a couple of commits, txs!

srmainwaring commented:
@rmackay9 it's looking good. I ran both the above cases for 45 deg and 90 deg `MOT_VEC_ANGLEMAX` - both ok. 

I noticed that the steering output for reverse is opposite when disarmed than when armed:

| Status | CH1_IN | CH3_IN | CH1_OUT | CH3_OUT | Note |
| --- | --- | --- | --- | --- | --- |
| Arm      | 1200 | 1700 | 1213 | 1680 | |
| Disarm | 1200 | 1700 | 1213 | 1500 | |
| Arm      | 1200 | 1300 | 1213 | 1320 | * |
| Disarm | 1200 | 1300 | 1787 | 1500 | * |
| Arm      | 1800 | 1700 | 1787 | 1680 | |
| Disarm | 1800 | 1700 | 1787 | 1500 | |
| Arm      | 1800 | 1300 | 1787 | 1320 | ** |
| Disarm | 1800 | 1300 | 1213 | 1500 | ** |

The cases of interest are marked * and **.

Thanks for the credit on the commits!

rmackay9 commented:
@srmainwaring,

Thanks for final check.  I've looked at the difference in direction of the steering output when armed vs disarmed and I can see why it is happening - when disarmed, the throttle input is ignored and the steering is simply passed straight through.  I think this behaviour is OK though..when disarmed the user just wants to check that the steering is moving correctly, I'm not sure they would expect the direction to change based on the ignored throttle input.. it could be done either way I think but as-is is probably fine too.

Txs again I think I'll merge shortly.

IamPete1 reviewed (approved):
LGTM

rmackay9 commented:
Merged, thanks!
